src/render/src/encoder.rs

```diff
@@ -593,50 +595,19 @@ impl<'a, B: Backend, C> Encoder<'a, B, C>
     }
 
-    /*
-    fn draw_indexed<T>(&mut self, buf: &handle::Buffer<B, T>, ty: IndexType,
-                    slice: &slice::Slice<B>, base: VertexCount,
-                    instances: Option<command::InstanceParams>) {
-        self.access_info.buffer_read(buf.raw());
-        self.command_buffer.bind_index(self.handles.ref_buffer(buf.raw()).clone(), ty);
-        self.command_buffer.call_draw_indexed(slice.start, slice.end - slice.start, base, instances);
-    }
-
-    fn draw_slice(&mut self, slice: &slice::Slice<B>, instances: Option<command::InstanceParams>) {
-        match slice.buffer {
-            slice::IndexBuffer::Auto => self.command_buffer.call_draw(
-                slice.start + slice.base_vertex, slice.end - slice.start, instances),
-            slice::IndexBuffer::Index16(ref buf) =>
-                self.draw_indexed(buf, IndexType::U16, slice, slice.base_vertex, instances),
-            slice::IndexBuffer::Index32(ref buf) =>
-                self.draw_indexed(buf, IndexType::U32, slice, slice.base_vertex, instances),
-        }
-    }
-
-    /// Draws a `slice::Slice` using a pipeline state object, and its matching `Data` structure.
-    pub fn draw<D: pso::PipelineData<B>>(&mut self, slice: &slice::Slice<B>,
-                pipeline: &pso::PipelineState<B, D::Meta>, user_data: &D)
+    pub fn draw<D>(
+        &mut self,
+        vertices: Range<VertexCount>,
+        pipeline: &D::Pipeline,
+        data: D
+    )
+        where D: pso::GraphicsPipelineData<B>
     {
-        let (pso, _) = self.handles.ref_pso(pipeline.get_handle());
-        //TODO: make `raw_data` a member to this struct, to re-use the heap allocation
-        self.raw_pso_data.clear();
-        user_data.bake_to(&mut self.raw_pso_data, pipeline.get_meta(), &mut self.handles, &mut self.access_info);
-        self.command_buffer.bind_pixel_targets(self.raw_pso_data.pixel_targets.clone());
-        self.command_buffer.bind_pipeline_state(pso.clone());
-        self.command_buffer.bind_vertex_buffers(self.raw_pso_data.vertex_buffers.clone());
-        self.command_buffer.set_ref_values(self.raw_pso_data.ref_values);
-        self.command_buffer.set_scissor(self.raw_pso_data.scissor);
-        self.command_buffer.bind_constant_buffers(&self.raw_pso_data.constant_buffers);
-        for &(location, value) in &self.raw_pso_data.global_constants {
-            self.command_buffer.bind_global_constant(location, value);
-        }
-        self.command_buffer.bind_unordered_views(&self.raw_pso_data.unordered_views);
-        //Note: it's important to bind RTV, DSV, and UAV before SRV
-        self.command_buffer.bind_resource_views(&self.raw_pso_data.resource_views);
-        self.command_buffer.bind_samplers(&self.raw_pso_data.samplers);
-        self.draw_slice(slice, slice.instances);
+        // TODO: instances
+        data.begin_renderpass(self, pipeline).draw(vertices, 0..1);
```

Comment:
starting and ending a render pas for each draw call is not perfect.
I think one way we cold make this nicer is to have encoders behave in the same way they do in Metal (it is a nice API, no matter what political negativity is around it):
  - have command buffers visible to the user
  - user creates a graphics encoder, which starts a render pass given the targets, and borrows the command buffer mutably
  - user submits a bunch of gaphics/draw commands
  - user drops the encoder, which ends the render pass and releases the command buffer borrow, which can now be submitted

cc @msiglreith

Reply:
I would be totally happy if we can get a Metal-ish API, but I also see why you went with the renderpass for each pipeline approach. Sounds like a hard problem due to required compatibility between framebuffer, pipeline and renderpass.
We might have to dynamically compile renderpasses (with single renderpass) and pipelines here to get this working or require the user to do it upfront or a mix of the both approaches.

Reply:
The way I see it:
  - user declares a render pass using a macro, like the other things. The information required for the macro is exactly what Vulkan considers to **not** be a part of render pass compatibility.
  - the "compatible" pieces would then be supplied at run-time during render pass creation by the user
  - the pipeline definition would then point to a render pass definition, thus statically enforcing which pipelines are going to be accepted in the graphics encoder corresponding to the given render pass definition (not the instance)

Does it make sense?

Reply:
Sounds reasonable even though I'm unsure if it's that easy to split the renderpass definition in compatible and non-compatible parts..

Reply:
It would be nice but I think it deserves a separate PR.

Reply:
that's fair